jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ContextProvider.java

```diff
@@ -589,56 +628,397 @@ else if (StringUtil.asciiStartsWithIgnoreCase(contextPath, "root-"))
         {
             int dash = contextPath.indexOf('-');
             String virtual = contextPath.substring(dash + 1);
-            context.setVirtualHosts(Arrays.asList(virtual.split(",")));
+            contextHandler.setVirtualHosts(Arrays.asList(virtual.split(",")));
             contextPath = "/";
         }
 
         // Ensure "/" is Prepended to all context paths.
         if (contextPath.charAt(0) != '/')
             contextPath = "/" + contextPath;
 
-        // Set the display name and context Path
-        context.setDisplayName(basename);
-        context.setContextPath(contextPath);
+        if (LOG.isDebugEnabled())
+            LOG.debug("ContextHandler {} initialized with displayName: {}", contextHandler, basename);
+        contextHandler.setDisplayName(basename);
+        if (LOG.isDebugEnabled())
+            LOG.debug("ContextHandler {} initialized with contextPath: {}", contextHandler, contextPath);
+        contextHandler.setContextPath(contextPath);
+    }
+
+    /**
+     * Apply the main deployable heuristics referenced in the main javadoc
+     * for this class.
+     *
+     * @param unit the unit of deployment to interrogate.
+     * @return the main deployable path
+     */
+    @Override
+    protected Path getMainDeploymentPath(Unit unit)
+    {
+        // XML always win.
+        List<Path> xmls = unit.getPaths().keySet().stream()
+            .filter(FileID::isXml)
+            .toList();
+        if (xmls.size() == 1)
+            return xmls.get(0);
+        else if (xmls.size() > 1)
+            throw new IllegalStateException("More than 1 XML for deployable " + asStringList(xmls));
+        // WAR files are next.
+        List<Path> wars = unit.getPaths().keySet().stream()
+            .filter(FileID::isWebArchive)
+            .toList();
+        if (wars.size() == 1)
+            return wars.get(0);
+        else if (wars.size() > 1)
+            throw new IllegalStateException("More than 1 WAR for deployable " + asStringList(wars));
+        // Directories next.
+        List<Path> dirs = unit.getPaths().keySet().stream()
+            .filter(Files::isDirectory)
+            .toList();
+        if (dirs.size() == 1)
+            return dirs.get(0);
+        if (dirs.size() > 1)
+            throw new IllegalStateException("More than 1 Directory for deployable " + asStringList(dirs));
+
+        throw new IllegalStateException("Unable to determine main deployable for " + unit);
```

Comment:
Is this really an ISE?   If we have just created `foo.d` or `foo.properties`, then we don't have enough to deploy anything, but I don't think it is an exceptional condition.  Perhaps a null return would be better, reflecting a unit that does not have a deployable path.

Reply:
Good point.
We need a testcase for this scenario.

What would you expect the result to be?
Kind feels like a "core" deployment, not a servlet deployment.

Reply:
just return null if there is no deployable file.